File: go/secret-handshake/secret_handshake.go
// Package secret contains methods to decode a secret handshake.
package secret

import "math"

var events = map[float64]string{
	1: "wink",
	2: "double blink",
	4: "close your eyes",
	8: "jump",
}

// Handshake decodes the input code into a list of handshake movements.
func Handshake(code uint) []string {
	var i uint
	movements := make([]string, 0)

	for i = 0; i <= 3; i++ {
		if (1<<i)&code > 0 {
			movements = append(movements, events[math.Pow(float64(2), float64(i))])
		}
	}

	if 16&code > 0 {
		return reverse(movements)
	}

	return movements
}

func reverse(strings []string) []string {
	for i, j := 0, len(strings)-1; i < j; i, j = i+1, j-1 {
		strings[i], strings[j] = strings[j], strings[i]
	}

	return strings
}

File: c/secret-handshake/src/secret_handshake.c
#include "secret_handshake.h"
#include <stdlib.h>

size_t calc_needed_size(size_t number);
const char **reverse(const char **string_array);
void swap(const char **string_array, int index1, int index2);

const int values[4] = {1, 2, 4, 8};
const char *moves[4] = {"wink", "double blink", "close your eyes", "jump"};

const char **commands(size_t number)
{
	const char **secret_handshake = calloc(calc_needed_size(number), sizeof(char *));
	int moveCounter = 0, i;

	for (i = 0; i < 4; i++)
		if ((number & values[i]) > 0)
			secret_handshake[moveCounter++] = moves[i];

	if ((number & 16) > 0) 
		return reverse(secret_handshake);

	return secret_handshake;
}

size_t calc_needed_size(size_t number)
{
	return (number & 1) + ((number >> 1) & 1) + ((number >> 2) & 1) + ((number >> 3) & 1);
}

const char **reverse(const char **string_array)
{
	int i, j;

	for (i = 0, j = 3; i < j; i++, j--)
		if (string_array[j])
			swap(string_array, i, j);
		else
			i--;

	return string_array;
}

void swap(const char **string_array, int index1, int index2)
{
	const char *temp = string_array[index1];
	string_array[index1] = string_array[index2];
	string_array[index2] = temp;
}
